workshop/stringBasic/StringBasic.cpp

```diff
@@ -0,0 +1,245 @@
+
+#include "stringBasic.h"
+
+StringBasic::StringBasic() :
+    m_string(nullptr),
+    m_lenght(0),
+    m_size(0)
+{}
+
+
+StringBasic::StringBasic(const char* const string) :
+    m_string(nullptr),
+    m_lenght(strlen(string)),
+    m_size(strlen(string))
+{
+    m_size *= SIZE_SCALE;
+    m_string = new char[m_size];
+    std::memset(m_string, 0, m_size);
+    std::memcpy(m_string, string, m_lenght);
+}
+
+
+StringBasic::~StringBasic() {
+    if (m_string != nullptr) {
+        try {
+            delete[] m_string;
+        }
+        catch (...) {
+        }
+        m_string = nullptr;
+    }
+}
+
+
+StringBasic::StringBasic(const StringBasic& other) :
+    m_size(other.m_size),
+    m_lenght(other.m_lenght),
+    m_string(nullptr)
+{
+    m_string = new char[m_size];
+    std::memset(m_string, 0, m_size);
+    std::memcpy(m_string, other.m_string, m_lenght);
+}
+
+
+StringBasic& StringBasic::operator=(const StringBasic& other) {
+    if (m_string != nullptr) {
+        try {
+            delete[] m_string;
+        }
+        catch (...) {
+        }
+    }
+
+    *this = StringBasic(other);
+    return *this;
+}
+
+
+StringBasic::StringBasic(StringBasic&& other) noexcept :
+    m_size(other.m_size),
+    m_lenght(other.m_lenght),
+    m_string(other.m_string)
+{
+    other.m_string = nullptr;
+    other.m_lenght = 0;
+    other.m_size = 0;
+}
+
+
+StringBasic& StringBasic::operator=(StringBasic&& other) noexcept {
+    if (m_string != nullptr) {
+        try {
+            delete[] m_string;
+        }
+        catch (...) {
+        }
+    }
+
+    other.m_string = nullptr;
+    *this = std::move(other);
+    return *this;
+}
+
+
+const unsigned long StringBasic::getLenght() const noexcept {
+    return m_lenght;
+}
+
+
+const char StringBasic::operator[](std::size_t index) const{
+    if (index >= m_lenght) {
+        throw std::out_of_range("invalid index exception");
+    }
+
+    return m_string[index];
+}
+
+
+char& StringBasic::operator[](std::size_t index) {
+    if (index >= m_lenght) {
+        throw std::out_of_range("invalid index exception");
+    }
+
+    return m_string[index];
+}
+
+
+std::ostream& operator<<(std::ostream& strm, const StringBasic& str) noexcept {
+    for (std::size_t i = 0; i < str.m_lenght; i++) {
+        strm << str.m_string[i];
+    }
+    return strm;
+}
+
+
+StringBasic& StringBasic::operator+=(const StringBasic& rhs) {
+    add_string(rhs.m_string);
+    return *this;
+}
+
+
+StringBasic& StringBasic::operator+=(const char& rhs) {
+    add_char(rhs);
+    return *this;
+}
+
+
+StringBasic& StringBasic::operator+=(const char* const rhs) {
+    add_string(rhs);
+    return *this;
+}
+
+
+StringBasic StringBasic::operator+(const StringBasic& rhs) {
+    StringBasic string = *this;
+    string.add_string(rhs.m_string);
+    return string;
+}
+
+
+StringBasic StringBasic::operator+(const char* const rhs) {
+    StringBasic string = *this;
+    string.add_string(rhs);
+    return string;
```

Comment:
This is fine due to RVO, but if you're not familiar with the concept i'd suggest reading about it and thinking what we can do to optimize this code in case RVO isn't being performed.